Add tensor parallelism related mappings

hyunwoongko opened the pull request:
Discussed in https://github.com/huggingface/transformers/issues/13690

In addition, I also created some helper functions. 
cc @RezaYazdaniAminabadi @lucasleesw 

## Reviewers
@stas00 @jaketae @siddk

hyunwoongko committed:
Add tensor parallelism related mappings

hyunwoongko committed:
Modify 'transformers' to '..'

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15360). All of your documentation changes will be reflected on that endpoint.

hyunwoongko commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -15,6 +15,154 @@
+
+# ie. nn.Linear(3 * dim, dim)
+# only defined the models that have fused attention layer.
+FUSED_ATTENTION_MAPPING = {
```
Currently, there are only two models that use this method in the transformers.

hyunwoongko commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -15,6 +15,154 @@
+
+# ie. nn.Linear(out_dim, in_dim) or Conv1D()
+# only defined the models that have reversed parameters.
+REVERSED_PARAM_MAPPING = {
```
Currently, there are only two models that use this method in the transformers.

hyunwoongko commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -15,6 +15,154 @@
+}
+
+# All the mapping for tensor parallelism
+TENSOR_PARALLEL_MAPPING = {
```
I wrote the mappings of the two most representative models. Let's expand gradually.

hyunwoongko committed:
Fix circular import problem

hyunwoongko commented on src/transformers/utils/tensor_parallel_utils.py:
```diff
@@ -0,0 +1,147 @@
+
+# ie. nn.Linear(3 * dim, dim)
+# only defined the models that have fused attention layer.
+FUSED_ATTENTION_MAPPING = {
```
I have already done tensor parallelization of 70 models in parallelformers, there are only two models that use this method in the transformers now.

hyunwoongko replied:
- in the normal cases
![image](https://user-images.githubusercontent.com/38183241/151268636-82903ec5-4dc2-4f24-9557-c105b401d226.png)

- this case
![image](https://user-images.githubusercontent.com/38183241/151268647-4090cdeb-48ff-4e82-8858-ce45260739c6.png)

hyunwoongko commented on src/transformers/utils/tensor_parallel_utils.py:
```diff
@@ -0,0 +1,147 @@
+
+# ie. nn.Linear(out_dim, in_dim) or Conv1D()
+# only defined the models that have reversed parameters.
+REVERSED_PARAM_MAPPING = {
```
I have already done tensor parallelization of 70 models in parallelformers, there are only two models that use this method in the transformers now.

hyunwoongko commented on src/transformers/utils/tensor_parallel_utils.py:
```diff
@@ -0,0 +1,147 @@
+}
+
+# All the mapping for tensor parallelism
+TENSOR_PARALLEL_MAPPING = {
```
I wrote the mappings of the two most representative models. Let's expand gradually.

lucasleesw replied:
Let me add a mapping for t5:
```
'T5PreTrinedModel': {
        'column_parallel':['Attention.q', 'Attention.k', 'Attention.v', 'DenseReluDense.wi', 'relative_attention_bias'],
        'row_parallel':['Attention.o', 'DenseReluDense.wo'],
        'update_attrs':['n_heads', 'inner_dim']
  }
```
But the parameter `relative_attention_bias` is not in Conv1D layer or a nn.Linear layer, but its weight should changed in TP as well. Do we need create a new kind of mapping for it?

hyunwoongko replied:
Exceptions of this kind are increasing day by day. So I designed it like this.
https://github.com/tunib-ai/parallelformers/blob/main/parallelformers/policies/t5.py

hyunwoongko replied:
The most safe way is using an object that describe each module like this.
https://github.com/tunib-ai/parallelformers/blob/main/parallelformers/policies/base/policy.py#L23

hyunwoongko replied:
So, @lucasleesw As @stas00 said, Let's consider adding a dictionary for exception cases. Creating objects for every model is very tiring. Let's create a dictionary like ``LAYER REPLACEMENT`` and define there only for anomalous cases like T5. Does it make sense?

stas00 replied:
Also some models will have an optional fused layernorm, e.g. very possibly with t5.

I've already mostly integrated apex.normalization.FusedRMSNorm in t5 https://github.com/huggingface/transformers/pull/14656 but for some reason it's not any faster, so need to find time to investigate why.

hyunwoongko replied:
@stas00 LayerNorm shouldn't be a problem since it doesn't affect tensor parallelization. Thanks for letting me know though.

stas00 replied:
oh, cool, I misread then that somewhere in this PR you mentioned layernorm. but it was attention! 

I guess it'd be more important to know for kernel fusion discussion.

hyunwoongko replied:
I added T5 and `no_replacement` element like this. this is optional. does it make sense?

hyunwoongko replied:
```python
TENSOR_PARALLEL_MAPPING = {
    BertPreTrainedModel: {
        "col": ["query", "key", "value", "intermediate.dense"],
        "row": ["output.dense"],
        "update": ["num_attention_heads", "all_head_size"],
    },
    GPT2PreTrainedModel: {
        "col": ["c_attn", "q_attn", "c_fc"],
        "row": ["c_proj"],
        "update": ["embed_dim", "split_size", "num_heads"],
    },
    T5PreTrainedModel: {
        "col": ["Attention.q", "Attention.k", "Attention.v", "DenseReluDense.wi"],
        "row": ["Attention.o", "DenseReluDense.wo"],
        "row_no_replacement": ["relative_attention_bias"],
        "update": ["d_model", "n_heads", "inner_dim"],
    },
}
```

hyunwoongko commented:
@stas00 Plus, In my opinion, this mapping + module replacement is the easiest to implement and the most extensible. However, there were three models (ProphetNet, BigBird, SqueezeBert) that could not be applied in this way. see: https://github.com/tunib-ai/parallelformers/blob/main/FAQ.md#q-why-are-some-models-not-supported

stas00 commented:
I think it's perfectly fine to design a way to automate most of the derivations if it works in most cases and leave the special cases to be that - special cases. Which can be taken care of later.

If you have ever used Perl, its motto: "make the easy things easy, and the hard things possible" and it lived up to that lofty goal. So if we can follow a similar principal here, then it's great! 

that's is let's not replicate data that doesn't need to replicated if it can be had automatically, and make it possible to overcome exceptions by providing additional data when it can't be auto-derived.

Does it make sense?

hyunwoongko commented:
@stas00 Yes. That will be fine. Let's proceed by adding mappings for exception cases. For example, most models match well, but if a few models have exceptions, it is easiest to solve the problem by constructing an additional dictionary for only those exception cases. The `REVERSED PARAMS` and `FUSED ATTENTION` maps I wrote are examples. These cases are not difficult to define because there are only two of them across transformers. @lucasleesw Let's design it to support as many models as possible with as little effort as possible. That would be the best.

hyunwoongko committed:
Add T5

hyunwoongko commented:
```python
ValueError: The following files have docstrings written in rst:
- src/transformers/utils/tensor_parallel_utils.py
To fix this run `doc-builder convert path_to_py_file` after installing `doc-builder`
(`pip install git+https://github.com/huggingface/doc-builder`)
make: *** [Makefile:40: repo-consistency] Error 1
```

@stas00 Can I solve this problem?
I think it's about documentation.

hyunwoongko committed:
Add copy mapping

hyunwoongko commented on src/transformers/utils/tensor_parallel_utils.py:
```diff
@@ -0,0 +1,196 @@
+        "row_no_replacement": ["relative_attention_bias"],
+        "update": ["d_model", "n_heads", "inner_dim"],
+    },
+    RobertaPreTrainedModel: BertPreTrainedModel,
```
I added the `copy_mapping` because there are some models that have same mapping in the Transformers.
In this case, we don't need to copy the mapping.

hyunwoongko committed:
Style code

hyunwoongko committed:
remove transformers dependancy

stas00 commented:
> ```python
> ValueError: The following files have docstrings written in rst:
> - src/transformers/utils/tensor_parallel_utils.py
> To fix this run `doc-builder convert path_to_py_file` after installing `doc-builder`
> (`pip install git+https://github.com/huggingface/doc-builder`)
> make: *** [Makefile:40: repo-consistency] Error 1
> ```
> 
> @stas00 @jaketae @siddk Can you help me? I think it's about documentation.
> 
> ```python
>  hyunwoongko  ~/Github/transformers   master ±  doc-builder convert path_to_py_file
> Traceback (most recent call last):
>   File "/Library/Frameworks/Python.framework/Versions/3.7/bin/doc-builder", line 8, in <module>
>     sys.exit(main())
>   File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/doc_builder/commands/doc_builder_cli.py", line 39, in main
>     args.func(args)
>   File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/doc_builder/commands/convert_doc_file.py", line 114, in convert_command
>     raise ValueError(f"This script only converts rst files. Got {source_file}.")
> ValueError: This script only converts rst files. Got /Users/hyunwoongko/Github/transformers/path_to_py_file.
> ```
> 
> Do I have to create a new rst file?

we switched from .rst to .mdx some weeks back. Please See the updated docs tree in master.

hyunwoongko committed:
Modify structure of tp mapping

hyunwoongko commented:
I modified tensor parallel mapping. I am sure this is the best structure. 

In conclusion, the list is not a suitable structure to represent this mapping because various parameter states exist, such as fusion and reverse, and they have many combinations. (fusion means the fused QKV parameters like linear(3 * dim, dim), not the kernel fusion)

“col_no_fuse”, “col_no_fuse_no_reverse”, “col_no_replace_no_fuse”, … I think this is not we want.
For example, it may or may not be reversed while being fused. Therefore, it is better to represent them as objects, and if we set the default values of attributes to the most common value, we can reduce the time required to create a mapping.

And I added many utility functions to search these states like:

- `mapping.get_fusion_degree(model, parameter)`
- `mapping.is_reversed(model, parameter)`
- `mapping.is_fused(model, parameter)`
- `mapping.update_attributes(model)`
- ...

hyunwoongko committed:
remove all the docstring

hyunwoongko committed:
Add code

hyunwoongko committed:
Fix bug

hyunwoongko committed:
Style code

hyunwoongko committed:
fuse to combined  qkv

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,176 @@
+            for elem in mapping:
+                if elem.code in self.__MAPPING__[cls]:
+                    self.__MAPPING__[cls][elem.code].append(elem)
+                else:
```
How about using a `defaultdict` in line 96, like 

```python
# from collections import defaultdict

self.__MAPPING__ = {cls: defaultdict(list) for cls in cache_tp_mapping}
```

so that we can remove the `else` here?

hyunwoongko replied:
Looks good to me !

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,176 @@
+            return mapping["Update"]
+
+    def search(self, model, param_name):
+        mapping = self.get_mapping(model)
```
Do we have to consider the case when`mapping` is `None`?

hyunwoongko replied:
I think It's good.

hyunwoongko replied:
It seems better to raise an assertion in `get_mapping` in the notice that the model is not currently supported. so we can remove the `None` state. How about you?

jaketae replied:
I agree Having to deal with `None` in other functions seems repetitive/bug-prone.

hyunwoongko replied:
In other methods, if None, it would be better to leave it open for other actions. I think it would be better to apply it only to this method.

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,176 @@
+            return bigger // smaller
+        return 1
+
+    def is_reversed_param(self, model, param_name):
```
Nit: For the `is_X` functions, I think it would be helpful to type annotate it with `def is_X() -> bool:` and add `return False` by default.

hyunwoongko replied:
Originally, all methods have docstring, but some errors related to documentation blocked the CI. Could you help me?

jaketae commented:
@hyunwoongko Apologies for the delayed review. I only have a few minor comments, and I can work on them if you'd like. Overall, I agree with the design choices you made. Thanks again for the PR!

hyunwoongko committed:
Fix bug of mapping

hyunwoongko committed:
Add GPTNeo, GPTJ

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+                    copy_elem.name = name
+                    cache_tp_mapping[cls].append(copy_elem)
+
+        self.__MAPPING__ = {cls: {} for cls in cache_tp_mapping}
```
```suggestion
        self.__MAPPING__ = {cls: defaultdict(list) for cls in cache_tp_mapping}
```

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+            for elem in mapping:
+                if elem.code in self.__MAPPING__[cls]:
+                    self.__MAPPING__[cls][elem.code].append(elem)
+                else:
```
```suggestion
```

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+
+        for cls, mapping in cache_tp_mapping.items():
+            for elem in mapping:
+                if elem.code in self.__MAPPING__[cls]:
+                    self.__MAPPING__[cls][elem.code].append(elem)
+                else:
+                    self.__MAPPING__[cls][elem.code] = [elem]
```
```suggestion
```

jaketae replied:
```suggestion
                self.__MAPPING__[cls][elem.code].append(elem)
```

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
-
+import copy
+import importlib
 from math import ceil
```
```suggestion
from collections import defaultdict
from math import ceil
```

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+            return mapping["Update"]
+
+    def search(self, model, param_name):
+        mapping = self.get_mapping(model)
```
```suggestion
        mapping = self.get_mapping(model)
        if mapping is None:
            raise ValueError(f"{model} does not support tensor parallelism.")
```

jaketae reviewed (commented):
I left committable review comments. Feel free to accept/reject/modify as you see fit, as they are more stylistic/nit. LGTM!

jaketae commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+    def get_mapping(self, model):
+        for cls, mapping in self.__MAPPING__.items():
+            if isinstance(model, cls):
+                return mapping
```
If we use `defaultdict`, we might want to re-wrap `mapping` around `dict` if we want to prevent `mapping["Row"]` from returning an empty list, for instance.

```suggestion
                return mapping
```

jaketae commented:
I think it would be good to have a basic test script. Here is a very basic sketch of what I had in mind.

```python
import unittest

from transformers import AutoModel
from transformers.utils.model_parallel_utils import TPMapping, Col, Row, Update


class TPMappingTest(unittest.TestCase):
    def setUp(self):
        self.tp_mapping = TPMapping()

    def test_bert(self):
        model = AutoModel.from_pretrained(f"hf-internal-testing/tiny-random-bert")
        self.assertDictEqual(
            self.tp_mapping.get_mapping(model),
            {
                'Col': [Col("query"), Col("key"), Col("value"), Col("intermediate.dense")],
                'Row': Row("output.dense"),
                'Update': [Update("num_attention_heads", "all_head_size")],
            }
        )

    # def test_some_other_model(self):
        # pass
```

At the moment, this test fails, and I'm assuming that the dictionary in the `self.assertDictEqual` statement is wrong. What would be the proper way to check for the output of `get_mapping()` @hyunwoongko ?

hyunwoongko committed:
Update src/transformers/utils/model_parallel_utils.py

Co-authored-by: Jake Tae <jaesungtae@gmail.com>

hyunwoongko committed:
Update src/transformers/utils/model_parallel_utils.py

Co-authored-by: Jake Tae <jaesungtae@gmail.com>

hyunwoongko committed:
Update src/transformers/utils/model_parallel_utils.py

Co-authored-by: Jake Tae <jaesungtae@gmail.com>

hyunwoongko committed:
Update src/transformers/utils/model_parallel_utils.py

Co-authored-by: Jake Tae <jaesungtae@gmail.com>

hyunwoongko committed:
Update src/transformers/utils/model_parallel_utils.py

Co-authored-by: Jake Tae <jaesungtae@gmail.com>

jaketae commented:
Hello @stas00, this PR is ready for review. Could you kindly take a look when time permits?

For context, this PR provides a mapping between model names and parameters that will enable tensor parallelism. While this PR was introduced as part of efforts to integrate Oslo-enabled 3D parallelism to transformers, the map was constructed in a way that is framework agnostic, i.e., it could be used by Oslo, DeepSpeed, and the likes.

cc @LysandreJik @sgugger

hyunwoongko commented:
@l-yohai If it works well after adding it to OSLO and testing it, let's continue to PR and add models here. I hope to support almost all models as soon as possible.

sgugger commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
 from math import ceil
 
 
+class TPInfo(object):
```
We need a docstring here explaining what this class does.

Also, the name is unclear to someone who is not familiar with all the model parallelism jargon. Let's expand to `TensorParallelismInfo` as we usually use descriptive names in Transformers.

sgugger commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
 
 
+class TPInfo(object):
+    def __init__(
+        self,
+        *name,
+        combined_qkv: bool = False,
+        reverse: bool = False,
+    ):
```
```suggestion
    def __init__(self, *name, combined_qkv: bool = False, reverse: bool = False):
```
This fits in one line (our char limit is 119).

sgugger commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+Update = type("UPDATE", (TPInfo,), {"code": "Update"})
+
+
+class TPMapping(object):
```
And here `TensorParallelismMapping`. We also need a docstring to explain what this does and how to expand it.

sgugger reviewed (approved):
Thanks for working on this. I've left a couple of suggestions regarding naming, as we usually go for descriptive variable names in Transformers.

We would also need docstrings in the classes introduced and their methods, or comments, before we merge this, so the code is documented and we can easily maintain it.

hyunwoongko commented:
@sgugger Thanks for fixing code. I'll apply these suggestion to code.
 
@jaketae I'll apply some suggestions, Could you help me adding docstrings?

stas00 commented:
I think that any such new mappings should have an application and tests with it, so I don't think this is a good addition to transformers the way it's presented at the moment as it could lead to dead code. 

Typically any new features always come with application, tests and documentation. This PR has none of that.

sgugger commented:
Oh I'm sorry, I thought this what part of a larger plan you had vetted together @stas00 and @jaketae . So indeed we want to make sure there is a tested application before merging.

sgugger reviewed (changes requested):
Oh I'm sorry, I thought this what part of a larger plan you had vetted together @stas00 and @jaketae . So indeed we want to make sure there is a tested application before merging.

hyunwoongko commented:
Then, I think I will have to add this when I add OSLO.

hyunwoongko commented:
closing the PR.

stas00 commented:
do you not want to first integrate the suggestions by Sylvain so that when added in the new PR it won't need to be done second time?

hyunwoongko commented:
@stas00 
> do you not want to first integrate the suggestions by Sylvain so that when added in the new PR it won't need to be done second time?

I will.

hyunwoongko commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+    @staticmethod
+    def _load_class_by_model_name(model_name):
+        transformers = importlib.import_module("transformers")
+        cls = getattr(transformers, f"{model_name}PreTrainedModel", None)
+        if cls is None:
+            cls = getattr(transformers, f"{model_name}PretrainedModel", None)
```
Can you let me know what version this was applied from?
The users of the non-existent version of the `BartPreTrainedModel` of transformers still need this.

To add a little more, the transformers will also have this mapping, but the Oslo will have it internally as well. This is because this mapping class does not exist for users of lower versions of the transformers. Users with lower versions will use the mappings inside OSLO and users with higher versions will use the mappings from transformers. 

This check is not required by transformers, but still required by OSLO.

hyunwoongko commented on src/transformers/utils/model_parallel_utils.py:
```diff
@@ -12,10 +12,186 @@
+        cls = getattr(transformers, f"{model_name}PreTrainedModel", None)
+        if cls is None:
+            cls = getattr(transformers, f"{model_name}PretrainedModel", None)
+        assert cls is not None, f"Can not import the model named {cls}."
```
Thanks for correction ! I think `if cls is None` is more correct. I'll apply your suggestion.
